packages/cursorless-engine/src/util/allocateHats/allocateHats.ts

```diff
@@ -107,10 +113,12 @@ export function allocateHats(
       }
 
       // Remove the hat we chose from consideration for lower ranked tokens
-      delete graphemeRemainingHatCandidates.get(chosenHat.grapheme.text)[
-        chosenHat.style
-      ];
-
+      graphemeRemainingHatCandidates.set(
+        chosenHat.grapheme.text,
+        graphemeRemainingHatCandidates
+          .get(chosenHat.grapheme.text)
+          .filter((style) => style !== chosenHat.style),
+      );
```

Comment:
Wow is this really faster than what I had? Performance tuning can be quite counterintuitive 😅

Reply:
Ah, here's a reference for you: https://v8.dev/blog/fast-properties

Reply:
yikes 😅. thanks for the reference 🙏

Reply:
@auscompgeek fabulous reference. Thank you!

The slowest part now by far is this line:

```ts
      candidates.push({
        grapheme,
        style,
        penalty: enabledHatStyles[style].penalty,
      });
```

Any tips for that? I can't tell whether it is object allocation, array growth, or what. Nor would I know what to do with that. (In Go, I would pre-allocate the slice and use an object pool.)

Any ideas? :)

Reply:
Maybe split it into 3 lines so profiling can tell you which piece is the problem? (Ie first get penalty, then construct object, then push)

Reply:
Oh right, now I remember I did that. it is all in constructing the object.

Reply:
Oh interesting. I wonder if there's a way to make that object construction faster. Any ideas @auscompgeek

Reply:
I can't think of any good ideas other than turning this into WebAssembly instead, sorry 😅

Reply:
Ha yeah fair. Rewrite in rust / zig / c into wasm would prob be much faster. But would hurt ability to easily experiment with this code I think

As a less dramatic change, it appears switching to a function prototype might speed things up. See eg https://www.measurethat.net/Benchmarks/Show/19556/0/es6-class-vs-prototype-vs-object-literal

We could also potentially switch to an array, tho that would make all the code that consumes these objects a lot less readable

Reply:
> I can't think of any good ideas other than turning this into WebAssembly instead, sorry 😅
> [...] Rewrite in rust [edit: or Go] [...]

omg don't tempt me

> it appears switching to a function prototype might speed things up

Using a class cuts about 5%, with fairly minimal readability impact (diff below). But it's hard to get excited about that. Using a function prototype shaves off a few more percent, but makes typescript grouchy. I think this is probably not worth the code.

> We could also potentially switch to an array

At some point I experimented with an array. The diff was prohibitive--we'd be better off with webasm.

---

```diff
diff --git a/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts b/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts
index 09bb49fc..ab3c5f92 100644
--- a/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts
+++ b/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts
@@ -140,6 +140,18 @@ function getTokenOldHatMap(oldTokenHats: readonly TokenHat[]) {
   return tokenOldHatMap;
 }
 
+class ConcreteHatCandidate implements HatCandidate {
+  grapheme: Grapheme;
+  style: HatStyleName;
+  penalty: number;
+
+  constructor(grapheme: Grapheme, style: HatStyleName, penalty: number) {
+    this.grapheme = grapheme;
+    this.style = style;
+    this.penalty = penalty;
+  }
+}
+
 function getTokenRemainingHatCandidates(
   tokenGraphemeSplitter: TokenGraphemeSplitter,
   token: Token,
@@ -156,11 +168,13 @@ function getTokenRemainingHatCandidates(
       // Allocating and pushing all of these objects is
       // the single most expensive thing in hat allocation.
       // Please pay attention to performance when modifying this code.
-      candidates.push({
-        grapheme,
-        style,
-        penalty: enabledHatStyles[style].penalty,
-      });
+      candidates.push(
+        new ConcreteHatCandidate(
+          grapheme,
+          style,
+          enabledHatStyles[style].penalty,
+        ),
+      );
     }
   }
   return candidates;
```